s3-vfs/src/main/java/org/pentaho/s3common/S3CommonFileObject.java

```diff
@@ -411,6 +413,23 @@ protected void doRename( FileObject newFile ) throws Exception {
     delete();
   }
 
+  private void doFolderMove( FileObject sourceFolder, FileObject targetFolder ) throws FileSystemException {
+    logger.trace( "creating folder [{}]", targetFolder.getPublicURIString() );
+    targetFolder.createFolder();
+    FileObject[] children = sourceFolder.getChildren();
+    for ( FileObject child : children ) {
+      FileObject targetChild = targetFolder.resolveFile( child.getName().getBaseName() );
+      if ( child.isFolder() ) {
+        // logger.trace( "moving folder recursive [{}] -> [{}]", child.getPath(), targetChild.getPath() );
+        doFolderMove( child, targetChild );
+      } else if ( child.isFile() ) {
+        logger.trace( "moving file [{}] -> [{}]", child.getPublicURIString(), targetChild.getPublicURIString() );
+        child.moveTo( targetChild );
+      }
```

Comment:
does the order here matter? Like, do we need to create the target folders before moving the child files?

Reply:
No, I think it's only really necessary for empty folders